File: bot_core.py



import os
import time
import requests

from belief_state import BeliefState
from proof_detector import score_proof, extract_candidate_stance  # noqa: F401


MODEL_NAME = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"
API_URL = f"https://api-inference.huggingface.co/models/{MODEL_NAME}"

REASONING_TRIGGER_WORDS = [
    "yes", "change", "update", "evidence",
    "should", "relevant", "current", "published",
    "study", "meta", "analysis", "found", "support", "suggest",
]


def load_model():
    """
    Kept for interface compatibility with app.py (which calls load_model()
    once at startup). There's no local model to load anymore -- this just
    checks that an HF_TOKEN is present and returns placeholders.
    """
    token = os.environ.get("HF_TOKEN")
    if not token:
        raise RuntimeError(
            "HF_TOKEN environment variable is not set. "
            "Get a free token at huggingface.co -> Settings -> Access Tokens, "
            "then set it as an environment variable named HF_TOKEN."
        )
    print("Using Hugging Face Inference API (no local model load needed).", flush=True)
    return None, None  # (model, tokenizer) placeholders -- unused in this version


def _call_api(prompt, max_new_tokens=60, retries=3):
    """
    Calls the HF Inference API. Handles the common 'model is loading' cold
    start by retrying after a short wait, since free-tier models unload when
    idle and take 10-30s to spin back up on first use.
    """
    token = os.environ["HF_TOKEN"]
    headers = {"Authorization": f"Bearer {token}"}
    payload = {
        "inputs": prompt,
        "parameters": {
            "max_new_tokens": max_new_tokens,
            "do_sample": True,
            "top_p": 0.9,
            "temperature": 0.8,
            "return_full_text": False,
        },
    }

    for attempt in range(retries):
        response = requests.post(API_URL, headers=headers, json=payload, timeout=30)

        if response.status_code == 200:
            data = response.json()
            # Standard text-generation response shape: [{"generated_text": "..."}]
            if isinstance(data, list) and data and "generated_text" in data[0]:
                text = data[0]["generated_text"].strip()
                return text.split("\n")[0]
            return ""

        if response.status_code == 503:
            # Model is cold-starting on Hugging Face's side -- wait and retry.
            wait_s = 15
            print(f"Model loading on HF servers, retrying in {wait_s}s...", flush=True)
            time.sleep(wait_s)
            continue

        # Any other error: surface it clearly instead of failing silently.
        raise RuntimeError(f"HF Inference API error {response.status_code}: {response.text[:300]}")

    raise RuntimeError("HF Inference API did not respond successfully after retries.")


def build_reasoning_prompt(belief, proof_score, user_message):
    return (
        f"Current belief: {belief.stance}\n"
        f"Proof score of user message: {proof_score:.2f}\n"
        f"User said: {user_message}\n\n"
        f"A proof score above 0.6 means real evidence was given. "
        f"Should the belief change based on this? Reason briefly:\n"
    )


def build_reply_prompt(belief, reasoning, history, user_message):
    prompt = f"I believe {belief.stance}.\n\n"
    for speaker, text in history[-4:]:
        prompt += f"{speaker}: {text}\n"
    prompt += f"User: {user_message}\nBot: I"
    return prompt


def generate_reply(model, tokenizer, prompt, max_new_tokens=60):
    # `model` and `tokenizer` args are kept so call sites don't need to change,
    # but they're unused now -- generation happens via the API instead.
    return _call_api(prompt, max_new_tokens=max_new_tokens)


def extract_belief_from_evidence(model, tokenizer, user_message, old_stance):
    prompt = (
        f"The following message contains evidence for a new position:\n"
        f"{user_message}\n\n"
        f"Summarize the new belief being argued for in one short clean sentence:\n"
    )
    summary = generate_reply(model, tokenizer, prompt, max_new_tokens=30)
    return summary if summary else old_stance


def run_turn(model, tokenizer, belief: BeliefState, history: list, user_message: str):
    proof_score = score_proof(user_message)

    reasoning_prompt = build_reasoning_prompt(belief, proof_score, user_message)
    reasoning = generate_reply(model, tokenizer, reasoning_prompt, max_new_tokens=40)

    reasoning_says_change = any(word in reasoning.lower() for word in REASONING_TRIGGER_WORDS)
    candidate = extract_belief_from_evidence(model, tokenizer, user_message, belief.stance)
    changed = belief.consider_update(proof_score, candidate) if reasoning_says_change else False
    if not changed:
        belief.hold_firm()

    reply_prompt = build_reply_prompt(belief, reasoning, history, user_message)
    reply = generate_reply(model, tokenizer, reply_prompt, max_new_tokens=80)

    history.append(("User", user_message))
    history.append(("Bot", reply))

    debug = {
        "proof_score": proof_score,
        "reasoning": reasoning,
        "changed": changed,
        "stance": belief.stance,
    }
    return reply, debug

File: web/lib/bot_core.js
// Port of bot_core.py. Calls the Hugging Face free Inference API directly
// from the browser using a token the user supplies (kept in-memory only).

import { scoreProof } from "./proof_detector.js";

const MODEL_NAME = "TinyLlama/TinyLlama-1.1B-Chat-v1.0";
const API_URL = `https://router.huggingface.co/hf-inference/models/${MODEL_NAME}`;

const REASONING_TRIGGER_WORDS = [
  "yes", "change", "update", "evidence",
  "should", "relevant", "current", "published",
  "study", "meta", "analysis", "found", "support", "suggest",
];

async function callApi(token, prompt, maxNewTokens = 60, retries = 3, onRetry = null) {
  for (let attempt = 0; attempt < retries; attempt++) {
    const response = await fetch(API_URL, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        inputs: prompt,
        parameters: {
          max_new_tokens: maxNewTokens,
          do_sample: true,
          top_p: 0.9,
          temperature: 0.8,
          return_full_text: false,
        },
      }),
    });

    if (response.status === 200) {
      const data = await response.json();
      if (Array.isArray(data) && data.length && "generated_text" in data[0]) {
        const text = data[0].generated_text.trim();
        return text.split("\n")[0];
      }
      return "";
    }

    if (response.status === 503) {
      // Model cold-starting on HF's side. Wait and retry.
      const waitS = 15;
      if (onRetry) onRetry(waitS, attempt + 1, retries);
      await new Promise((res) => setTimeout(res, waitS * 1000));
      continue;
    }

    const bodyText = await response.text();
    throw new Error(`HF Inference API error ${response.status}: ${bodyText.slice(0, 300)}`);
  }

  throw new Error("HF Inference API did not respond successfully after retries.");
}

function buildReasoningPrompt(belief, proofScore, userMessage) {
  return (
    `Current belief: ${belief.stance}\n` +
    `Proof score of user message: ${proofScore.toFixed(2)}\n` +
    `User said: ${userMessage}\n\n` +
    `A proof score above 0.6 means real evidence was given. ` +
    `Should the belief change based on this? Reason briefly:\n`
  );
}

function buildReplyPrompt(belief, reasoning, history, userMessage) {
  let prompt = `I believe ${belief.stance}.\n\n`;
  for (const [speaker, text] of history.slice(-4)) {
    prompt += `${speaker}: ${text}\n`;
  }
  prompt += `User: ${userMessage}\nBot: I`;
  return prompt;
}

async function generateReply(token, prompt, maxNewTokens = 60, onRetry = null) {
  return callApi(token, prompt, maxNewTokens, 3, onRetry);
}

async function extractBeliefFromEvidence(token, userMessage, oldStance, onRetry = null) {
  const prompt =
    `The following message contains evidence for a new position:\n` +
    `${userMessage}\n\n` +
    `Summarize the new belief being argued for in one short clean sentence:\n`;
  const summary = await generateReply(token, prompt, 30, onRetry);
  return summary || oldStance;
}

export async function runTurn(token, belief, history, userMessage, onRetry = null) {
  const proofScore = scoreProof(userMessage);

  const reasoningPrompt = buildReasoningPrompt(belief, proofScore, userMessage);
  const reasoning = await generateReply(token, reasoningPrompt, 40, onRetry);

  const reasoningSaysChange = REASONING_TRIGGER_WORDS.some((w) =>
    reasoning.toLowerCase().includes(w)
  );
  const candidate = await extractBeliefFromEvidence(token, userMessage, belief.stance, onRetry);
  const changed = reasoningSaysChange ? belief.considerUpdate(proofScore, candidate) : false;
  if (!changed) belief.holdFirm();

  const replyPrompt = buildReplyPrompt(belief, reasoning, history, userMessage);
  const reply = await generateReply(token, replyPrompt, 80, onRetry);

  history.push(["User", userMessage]);
  history.push(["Bot", reply]);

  const debug = {
    proof_score: proofScore,
    reasoning,
    changed,
    stance: belief.stance,
    confidence: belief.confidence,
  };
  return { reply, debug };
}
